lib/runtime-core/src/state.rs

```diff
@@ -86,8 +109,84 @@ pub enum MachineValue {
     TwoHalves(Box<(MachineValue, MachineValue)>), // 32-bit values. TODO: optimize: add another type for inner "half" value to avoid boxing?
 }
 
+impl BorshSerialize for MachineValue {
+    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        match self {
+            MachineValue::Undefined => writer.write_all(&[0u8])?,
+            MachineValue::Vmctx => writer.write_all(&[1u8])?,
+            MachineValue::VmctxDeref(v) => {
+                writer.write(&[2u8])?;
```

Comment:
Hm, I *think* this is potentially buggy, and should be `write_all` instead. `write` can fail with `ErrorKind::Interrupted` and should be retried in that case. `write_all` does retry, while `write` by itself does not. The `ErrorKind::Interrupted ` is returned when a signal arrives while the thread is trying to write data. 

I don't know if we are actually affected by this or not, but better safe than sorry.

Reply:
Hmm I don't think there's a difference, given write_all's impl:
https://doc.rust-lang.org/stable/src/std/io/mod.rs.html#1442-1457

Reply:
`Err(ref e) if e.kind() == ErrorKind::Interrupted => {}` that's the line that makes the difference. As far as I understand, even a single-byte write might fail with `ErrorKind::Interrupted`. `write_all` will retry, while `write` will propagate this error. The correct behavior here is to retry.

Reply:
:+1 got it